### PR TITLE
[functorch] Add some C++ documentation

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84648
* #84642
* #84639
* #84638
* __->__ #84637

Clarify the purpose of many files.

Differential Revision: [D39315053](https://our.internmc.facebook.com/intern/diff/D39315053)

### CODE REVIEW

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39315053).

File: functorch/functorch/csrc/DynamicLayer.h [hunk 1]
```diff
@@ -18,12 +18,28 @@
...
 
-// Forward declared bc I am lazy
+// Forward declared
 namespace c10 { struct AutogradMetaInterface; }
 
 namespace at {
 namespace functorch {
 
+// This file contains the implementation of functorch's interpreter stack.
+// See NOTE: [functorch interpreter stack] first before reading on.
```
Comment (samdow) [inline functorch/functorch/csrc/DynamicLayer.h:28]: Is this note already in the code? I'm just seeing it referenced a lot of places without seeing it, so wanted to make sure it wasn't forgotten about

Comment (zou3519) [reply functorch/functorch/csrc/DynamicLayer.h:28]: Yes: https://github.com/pytorch/pytorch/blob/31ef8ddb8c4467f5b8698ef1eb9bb8bab7056855/functorch/functorch/csrc/Interpreter.h#L12

**Review by samdow — APPROVED** [review]

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39315053).

**Pull request closed without merging**